dalle2_pytorch/train_diffusion_prior.py

```diff
@@ -0,0 +1,207 @@
+import argparse
+import os
+from dalle2_pytorch import DiffusionPrior
+from embedding_reader import EmbeddingReader
+from dalle2_pytorch import DiffusionPriorNetwork
+import numpy as np
+import math
+from tqdm import tqdm
+import torch
+from torch import nn
+from torch.optim import AdamW
+import wandb
+os.environ["WANDB_SILENT"] = "true"
+
+def separate_weight_decayable_params(params):
+    no_wd_params = set([param for param in params if param.ndim < 2])
+    wd_params = set(params) - no_wd_params
+    return wd_params, no_wd_params
+
+def get_optimizer(params, lr = 3e-4, wd = 1e-1, filter_by_requires_grad = False):
+    if filter_by_requires_grad:
+        params = list(filter(lambda t: t.requires_grad, params))
+
+    params = set(params)
+    wd_params, no_wd_params = separate_weight_decayable_params(params)
+
+    param_groups = [
+        {'params': list(wd_params)},
+        {'params': list(no_wd_params), 'weight_decay': 0},
+    ]
+
+    return AdamW(param_groups, lr = lr, weight_decay = wd)
+
+def eval_model(model,device,image_reader,text_reader,start,end,batch_size,isValidation=True):
+    with torch.no_grad():
+        for emb_images,emb_text in zip(image_reader(batch_size=batch_size, start=start, end=end),
+                text_reader(batch_size=batch_size, start=start, end=end)):
+            emb_images = list(emb_images)
+            emb_text = list(emb_text)
+            emb_images[0] = torch.tensor(emb_images[0]).to(device)
+            emb_text[0] = torch.tensor(emb_text[0]).to(device)
+            model.eval()
+            loss = model(text_embed = emb_text[0],image_embed = emb_images[0])
+
+            # Log to wandb
+            if(isValidation):
+                wandb.log({"Validation mse": loss})
+            else:
+                wandb.log({"Test mse": loss})
+
```

Comment:
the loss type may not always be `mse`, and you can get the loss type identifier by `diffusion_prior.loss_type`